daisy/disk.go

```diff
@@ -176,23 +182,27 @@ func (dr *diskRegistry) deleteFn(res *Resource) DError {
 }
 
 // detachHelper marks s as the detacher between dName and iName.
-// Returns an error if dName and iName aren't attached or if the detacher doesn't depend on the attacher.
+// Returns an error if the detacher doesn't depend on the attacher.
 func (dr *diskRegistry) detachHelper(dName, iName string, s *Step) DError {
 	if dr.testDetachHelper != nil {
 		return dr.testDetachHelper(dName, iName, s)
 	}
 	pre := fmt.Sprintf("step %q cannot detach disk %q from instance %q", s.name, dName, iName)
+
 	var att *diskAttachment
 
 	if im, _ := dr.attachments[dName]; im == nil {
-		return Errf("%s: not attached", pre)
+		return nil
```

Comment:
I don't think this is what we want, I think we need these errors, what we don't have is registering resources the workflow did not create right? If we register instances and all their attached devices all this logic should still work?

Reply:
You are right, if we want to keep the original check, we need to register all attached devices. I can add that. However, "attach" doesn't have the corresponding check. Do you think we should check whether a "AttachDisk" is not attached to any instance?

Reply:
The point of all the validate steps is to catch any errors before running the workflow, instead of fully checking all attached resources, maybe check on the fly if this is not a daisy created resource? So if a non daisy resource is referenced in an AttachDisk step, lookup attached disks, add them to the registry, then run the normal check. Does that make sense?

Reply:
"if a non daisy resource is referenced in an AttachDisk step, lookup attached disks, add them to the registry" - that is what AttachDisk do today, by regUse->regURL.

The hard part is DetachDisk, which uses "DeviceName" orther than "Disk" as parameter, but actually ref to a disk. I have already added a "regUseDeviceName" to resolve this problem, collaborating with a deviceNameURLRgx. It did the same thing as you described above - if the referenced resource is not a daisy resource (aka it's a projects/xxx/xx/xx URL), check whether the corresponding instance has a device attached already. Please review it.